### PR TITLE
Migrate to TF 0.12.3

### LINKED ISSUE #7: aws: some default values are not properly imported
We've noticed that some default variables from Terraform or not directly imported via their importer method - some are already referenced as bug from the CLI importer -:

* [ ] EC2 Security groups https://github.com/terraform-providers/terraform-provider-aws/issues/2329
* [ ] Route53 zone - force_destroy field
* [ ] EC2 Instance root_device_block_termination
* [ ] IAM https://github.com/terraform-providers/terraform-provider-aws/issues/7859
* [ ] Cloudfront retain on delete https://github.com/terraform-providers/terraform-provider-aws/commit/b46ceca07742633e829b282bc2786e782083dafb
* [ ] RDS db_instance skip final snapshot https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_db_instance.go#L1642
* [ ] S3 bucket https://github.com/terraform-providers/terraform-provider-aws/issues/6193

### PR DESCRIPTION
**Pull request by xescugc**
Change all the workflow to work with the new 0.12 implementation of TF. This implementation uses the Plugins logic without using Plugins directly.

The HCL and State have been changed to work with the new format.

Closes #7

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - The `version` subcommand to show the actual build version
   ([Issue #24](https://github.com/cycloidio/terracognita/issues/24))
 
+### Changed
+
+- The Terraform version from 0.11 to 0.12 with all the implications (file formats)
+  ([PR #33](https://github.com/cycloidio/terracognita/pull/33))
```
Comment (xlr-8) [inline CHANGELOG.md]: Could make sense to drop a link to https://www.terraform.io/upgrade-guides/0-12.html

File: hcl/format_hcl.go [hunk 2]
```diff
@@ -9,29 +11,38 @@ var (
...
+					return []byte(fmt.Sprintf("%s", m))
+				}
+
+				return bytes.Replace(m, []byte(`= `), nil, 1)
+
+			},
+		},
+		{
+			// Add new lines before blocks
+			//match: regexp.MustCompile("\n(\t*)([\\w\\-_\\.]+\\s{)"),
```
Comment (xlr-8) [inline hcl/format_hcl.go]: Unecessary commented code

File: hcl/format_hcl.go [hunk 3]
```diff
@@ -45,11 +56,15 @@ var (
 	}
 )
 
-// FormatHCL format the hcl to have a better formatter thatn the default one
+// Format format the hcl to have a better formatter thatn the default one
```
Comment (xlr-8) [inline hcl/format_hcl.go]: than*
returned*

File: hcl/writer.go [hunk 4]
```diff
@@ -50,22 +50,38 @@ func (hclw *HCLWriter) Write(key string, value interface{}) error {
...
+	if _, ok := w.Config["resource"].(map[string]map[string]interface{})[keys[0]]; !ok {
+		w.Config["resource"].(map[string]map[string]interface{})[keys[0]] = make(map[string]interface{})
 	}
-	hclw.Config["resource"].(map[string]map[string]interface{})[keys[0]][name] = value
+	w.Config["resource"].(map[string]map[string]interface{})[keys[0]][name] = value
 
 	return nil
 }
 
+// Has checks if the given key it's already present or not
```
Comment (xlr-8) [inline hcl/writer.go]: is already*

File: hcl/writer.go [hunk 5]
```diff
@@ -50,22 +50,38 @@ func (hclw *HCLWriter) Write(key string, value interface{}) error {
...
+func (w *Writer) Has(key string) (bool, error) {
+	keys := strings.Split(key, ".")
+	if len(keys) != 2 || (keys[0] == "" || keys[1] == "") {
```
Comment (xlr-8) [inline hcl/writer.go]: Unecessary parenthesis

File: provider/import.go [hunk 6]
```diff
@@ -35,33 +36,43 @@ func Import(ctx context.Context, p Provider, hcl, tfstate writer.Writer, f *filt
...
-				err = r.HCL(hcl)
+			// In case there is more than one State to import
+			// we create a new slice with those elements and iterate
+			// over it
+			for _, r := range append([]Resource{re}, res...) {
+				err = util.RetryDefault(func() error { return r.Read(f) })
 				if err != nil {
-					return errors.Wrapf(err, "error while calculating the Config of resource %q", t)
+					cause := errors.Cause(err)
+					// All those ignored errors are types of errors to identify that something happend, but not that something whent wrong
```
Comment (xlr-8) [inline provider/import.go]: went*

File: provider/import.go [hunk 7]
```diff
@@ -35,33 +36,43 @@ func Import(ctx context.Context, p Provider, hcl, tfstate writer.Writer, f *filt
...
+					err = r.HCL(hcl)
+					if err != nil {
+						return errors.Wrapf(err, "error while calculating the Config of resource %q", t)
+					}
+				}
+
+				if tfstate != nil {
+					err = r.State(tfstate)
+					if err != nil {
+						return errors.Wrapf(err, "error while calculating the Satate of resource %q", t)
```
Comment (xlr-8) [inline provider/import.go]: state*

File: state/writer.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	w.Config[key] = r
+
+	return nil
+}
+
+// Has checks if the given key it's already present or not
+func (w *Writer) Has(key string) (bool, error) {
+	_, ok := w.Config[key]
+	return ok, nil
+}
```
Comment (xlr-8) [inline state/writer.go]: I'm not sure this `Has` method or the other one are used anywhere

Comment (xescugc) [reply state/writer.go]: https://github.com/cycloidio/terracognita/pull/33/files#diff-9a4396fc8b469f87eb63a29771b3380fR330 also on the HCL.

The `resource.go` was `collapsed` on my diff maybe you forgot to review it.

**Comment by xescugc** [comment]
Did the changes.

File: provider/resource.go [hunk 9]
```diff
@@ -89,45 +116,84 @@ func (r *resource) ID() string { return r.id }
...
-	if r.resourceType == "aws_iam_user_group_membership" {
-		gps := r.data.Get("groups").(*schema.Set).List()
-		var gpsKey string
-		for _, gp := range gps {
-			if gpsKey == "" {
-				gpsKey = gp.(string)
-			}
-			gpsKey = fmt.Sprintf("%s/%s", gpsKey, gp)
+	resources := make([]Resource, 0, len(newInstanceStates)-1)
+	// We assume that the first state it's the one for this Resource
```
Comment (xlr-8) [inline provider/resource.go]: is the one*

File: provider/resource.go [hunk 10]
```diff
@@ -89,45 +116,84 @@ func (r *resource) ID() string { return r.id }
...
 func (r *resource) Provider() Provider { return r.provider }
 
-func (r *resource) Read(f *filter.Filter) error {
-	// Retry if any error happen
-	err := util.RetryDefault(func() error {
-		return r.tfResource.Read(r.data, r.provider.TFClient())
-	})
+func (r *resource) ImportState() ([]Resource, error) {
+	// If it does not support import do not try
+	if r.TFResource().Importer == nil {
```
Comment (xlr-8) [inline provider/resource.go]: Could it make sense to log that as a Warning/Debug info?

**Comment by xlr-8** [comment]
The code generated doesn't quite seem to be valid for terraform, some blocks/quote and optional params are generating errors.

**Comment by xescugc** [comment]
Ok done, some of the issues we are having now are related to TF not working on some form and we have to patch it on top normally. I would say to merge this 0.11 -> 0.12 and then fix the other issues in separated PRs.

File: util/retry.go [hunk 11]
```diff
@@ -25,6 +26,13 @@ func Retry(rfn RetryFn, times int, interval time.Duration) error {
 		if times == 0 {
 			return err
 		}
+		// If the error it's from the stdlib we just continue with them
```
Comment (xlr-8) [inline util/retry.go]: is from*

File: aws/resources.go [hunk 12]
```diff
@@ -854,23 +800,9 @@ func iamUserGroupMemberships(ctx context.Context, a *aws, resourceType string, t
...
-		r, err := initializeResource(a, NoID, resourceType)
-		if err != nil {
-			return nil, err
-		}
-		err = r.Data().Set("user", un)
-		if err != nil {
-			return nil, err
-		}
-		err = r.Data().Set("groups", groupSet)
+		r, err := initializeResource(a, strings.Join(append([]string{un}, groupNames...), "/"), resourceType)
```
[the remark this answers is not in the document]
Comment (xlr-8) [reply aws/resources.go]: > each format could help

Not only this one

File: provider/resource.go [hunk 13]
```diff
@@ -512,6 +513,12 @@ func mergeFullConfig(cfgr *schema.ResourceData, sch map[string]*schema.Schema, k
 			}
 		}
 
+		// If it's a type map, we'll prefix the key with '=tc=' to know
+		// that it's an attribute so we have to keep the '=' when formatting
+		// the HCL
```
Comment (xlr-8) [inline provider/resource.go]: Good idea :)

**Pull request merged**